en/tutorials-and-examples/cms/articles-controller.rst

```diff
@@ -247,11 +249,289 @@ to :php:func:`Cake\\Routing\\Router::url()` function on the `API
+
+The ``$this->Form->control()`` method is used to create form elements
+of the same name. The first parameter tells CakePHP which field
+they correspond to, and the second parameter allows you to specify
+a wide array of options - in this case, the number of rows for the
+textarea. There's a bit of introspection and conventions used here. The
+``control()`` will output different form elements based on the model
+field specified, and use inflection to generate the label text. If you want to
+customize the label, input or any other aspect of the form controls you can.
+
+The ``$this->Form->end()`` closes the form. Outputting hidden inputs if
+CSRF/Form Tampering prevention is enabled.
+
+Now let's go back and update our **src/Template/Articles/index.ctp**
+view to include a new "Add Article" link. Before the ``<table>``, add
+the following line::
+
+    <?= $this->Html->link('Add Article', ['action' => 'add']) ?>
+
+Adding Simple Slug Generation
+=============================
+
+If we were to save an Article right now, saving would fail as we are not
+creating a slug attribute, and the column is ``NOT NULL``. Slug values are
+typically a URL safe version of an article's title. We can use the
+:ref:`beforeSave() callback <table-callbacks>` of the ORM to populate our slug::
+
+    // in src/Model/Table/ArticlesTable.php
+    use Cake\Utility\Text;
+
+    // Add the following method.
+
+    public function beforeSave($event, $entity, $options)
+    {
+        if ($entity->isNew() && !$entity->slug) {
+            $entity->slug = Text::slug($entity->title);
+        }
+
+        // This is temporary, and will be removed later
+        // when we build authentication out.
+        if (!$entity->user_id) {
+            $entity->user_id = 1;
+        }
+    }
+
+This code is simple, and doesn't take into account duplicate slugs. But we'll
+fix that later on.
+
+Add Edit Action
+===============
+
+Our application can now save articles, but we can't edit them. Lets rectify that
+now. Add the following action to your ``ArticlesController``::
+
+    // in src/Controller/ArticlesController.php
+
+    // Add the following method.
+
+    public function edit($id = null)
+    {
+        $article = $this->Articles->get($id);
+        if ($this->request->is(['post', 'put'])) {
+            $this->Articles->patchEntity($article, $this->request->getData());
+            if ($this->Articles->save($article)) {
+                $this->Flash->success(__('Your article has been updated.'));
+                return $this->redirect(['action' => 'index']);
+            }
+            $this->Flash->error(__('Unable to update your article.'));
+        }
+
+        $this->set('article', $article);
+    }
+
+This action first ensures that the user has tried to access an existing record.
+If they haven't passed in an ``$id`` parameter, or the article does not
+exist, we throw a ``NotFoundException`` for the CakePHP ErrorHandler to take
```

Comment:
Instead of "we throw", what about "`$this->Articles->get($id)` will throw"? Then it explains where the NotFoundException is coming from